Sources/PackageGraph/Pubgrub.swift

```diff
@@ -516,6 +562,30 @@ public final class PubgrubDependencyResolver<
     /// Skip updating containers while fetching them.
     private let skipUpdate: Bool
 
+    var saveTraceSteps = false
+    var traceSteps: [TraceStep] = []
+    func trace(
+        value: String,
+        type: String,
+        location: String,
+        cause: String?
+        ) {
+        guard saveTraceSteps else { return }
+        traceSteps.append(TraceStep(value: value,
```

Comment:
Instead of saving the trace steps, we should report them to the delegate of resolver. That way the client can control what to do when them (like live-logging). Another advantage is that we can have a partial log in case there is a crash or something.